File: lc/c/012.c
void tranpose(int **matrix, int matrixSize, int *matrixColSize) {
    for (int r = 0; r < matrixSize; ++r) {
        for (int c = r; c < matrixColSize[r]; ++c) {
            int temp = matrix[r][c];
            matrix[r][c] = matrix[c][r];
            matrix[c][r] = temp;
        }
    }
}

void reflect(int **matrix, int matrixSize, int *matrixColSize) {
    for (int r = 0; r < matrixSize; ++r) {
        for (int c = 0; c < matrixColSize[r]/2; ++c) {
            int temp = matrix[r][c];
            matrix[r][c] = matrix[r][matrixSize + ~c];
            matrix[r][matrixSize + ~c] = temp;
        }
    }
}

void rotate(int** matrix, int matrixSize, int* matrixColSize){
    tranpose(matrix, matrixSize, matrixColSize);
    reflect(matrix, matrixSize, matrixColSize);
}

File: lc/py/012.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        # Flip over diagonal
        def transpose():
            for r in range(len(matrix)):
                for c in range(r, len(matrix[r])):
                    matrix[r][c], matrix[c][r] = matrix[c][r], matrix[r][c]

        # Reflect over vertical
        def reflect():
            for r in range(len(matrix)):
                for c in range(len(matrix[r])//2):
                    matrix[r][c], matrix[r][~c] = matrix[r][~c], matrix[r][c]

        transpose()
        reflect()

# Alternative solution
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        n = len(matrix)
        for i in range(n//2): # for layers
            for j in range(i, n-i-1): # for column, -1 to exclude last corner
                (matrix[i][j], matrix[~j][i], matrix[~i][~j], matrix[j][~i]) = (matrix[~j][i], matrix[~i][~j], matrix[j][~i], matrix[i][j])
